nbconvert/preprocessors/tests/test_execute.py

```diff
@@ -40,7 +42,64 @@ def _normalize_base64(b64_text):
     except (ValueError, TypeError):
         return b64_text
 
-class TestExecute(PreprocessorTestsBase):
+
+def merge_dicts(first, second):
+    # Because this is annoying to do inline
+    outcome = {}
+    outcome.update(first)
+    outcome.update(second)
+    return outcome
+
+
+class ExecuteTestBase(PreprocessorTestsBase):
+    def build_preprocessor(self, opts):
+        """Make an instance of a preprocessor"""
+        preprocessor = ExecutePreprocessor()
+        preprocessor.enabled = True
+        for opt in opts:
+            setattr(preprocessor, opt, opts[opt])
+        # Perform some state setup that should probably be in the init
+        preprocessor._display_id_map = {}
+        preprocessor.widget_state = {}
+        preprocessor.widget_buffers = {}
+        return preprocessor
+
+    @staticmethod
+    def prepare_cell_mocks(*messages):
```

Comment:
I think this needs a good deal of explanation at a high level before this will be easy to follow. I don't think we've been using mock much up to this point and this makes all of these heavily dependent on understanding how that works. I'm going to stop here, not because I don't have reviews to give for code further down but because this is going to take a few times going over with greater understanding on each pass given the new approach this is introducing.